### PR TITLE
Make UIA in MS Word optional again, through an Advanced Settings category

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
None.

### Summary of the issue:
NVDA switched to using UI Automation to access Microsoft Word document controls by default for Office 2016/365, however Office's UIA support is still not 100% complete.
This should be left optional for 2019.1 at least.

### Description of how this pull request fixes the issue:
UIA is no longer used by default in MS Word. However, there is now a new Advanced Settings category in NVDA's settings dialog which contains a checkbox to turn this on for testing this feature out.
The Advanced Settings category also contains all other config options previously not in the GUI. these include caret movement timeout, and the Debug log categories.

### Testing performed:
Tested each option in the Advanced settings category to ensure that their setting is remembered when opening the Settings dialog again, and that the option is truly activated / deactivated in NVDA itself.

### Known issues with pull request:
Although there is a panel caption warning the user to be careful with these settings, it is not automatically read by NVDA when tabbing from the categories list into the settings themselves. This is similar to the Document Formatting settings caption which is also not read.

### Change log entry:
New features:
A new Advanced Settings category has been added to NVDA's Settings dialog, including an option to try out NVDA's new support for Microsoft Word via the Microsoft UI Automation API.

### CODE REVIEW

File: source/gui/settingsDialogs.py [hunk 1]
```diff
@@ -1928,6 +1928,67 @@ def onSave(self):
...
+
+		self.debugLogCheckboxes={}
+		for key in (
+			"hwIo",
+			"audioDucking",
+			"gui",
+			"louis",
+			"timeSinceInput",
+		):
+			checkbox=debugLogGroup.addItem(wx.CheckBox(self, label=key))
```
Comment (leonardder) [inline source/gui/settingsDialogs.py]: Have you considered making this a checkable list? You can use the keyboard layouts list as an example.

File: user_docs/en/userGuide.t2t [hunk 2]
```diff
@@ -1603,6 +1603,26 @@ This category contains the following options:
 ==== Recognition language ====[Win10OcrSettingsRecognitionLanguage]
 This combo box allows you to choose the language to be used for text recognition.
 
++++ Advanced Settings +++
+Warning! The settings in this category are for advanced users and may cause NVDA to not function correctly if configured in the wrong way.
+Only make changes to these settings if you are sure you know what you are doing or if you have been specifically instrructed to by an NVDA developer.
```
Comment (lukaszgo1) [inline user_docs/en/userGuide.t2t]: instrructed = instructed

File: user_docs/en/userGuide.t2t [hunk 3]
```diff
@@ -1603,6 +1603,26 @@ This category contains the following options:
 ==== Recognition language ====[Win10OcrSettingsRecognitionLanguage]
 This combo box allows you to choose the language to be used for text recognition.
 
++++ Advanced Settings +++
+Warning! The settings in this category are for advanced users and may cause NVDA to not function correctly if configured in the wrong way.
+Only make changes to these settings if you are sure you know what you are doing or if you have been specifically instructed to by an NVDA developer.
+
+==== Use UI automation to access Microsoft Word document controls when available ====
+When this option is enabled, NVDA will try to use  the Microsoft UI Automation accessibility api in order to fetch information from Microsoft Word document controls.
+this includes in Microsoft Word itself, and also the Microsoft Outlook message viewer and composer.
```
Comment (leonardder) [inline user_docs/en/userGuide.t2t]: Doesn't start with a capital T
```suggestion
This includes documents in Microsoft Word itself, and also the Microsoft Outlook message viewer and composer.
```

File: user_docs/en/userGuide.t2t [hunk 4]
```diff
@@ -1603,6 +1603,26 @@ This category contains the following options:
...
 This combo box allows you to choose the language to be used for text recognition.
 
++++ Advanced Settings +++
+Warning! The settings in this category are for advanced users and may cause NVDA to not function correctly if configured in the wrong way.
+Only make changes to these settings if you are sure you know what you are doing or if you have been specifically instructed to by an NVDA developer.
+
+==== Use UI automation to access Microsoft Word document controls when available ====
+When this option is enabled, NVDA will try to use  the Microsoft UI Automation accessibility api in order to fetch information from Microsoft Word document controls.
+this includes in Microsoft Word itself, and also the Microsoft Outlook message viewer and composer.
+ For the most recent versions of Microsoft Office 2016/365 running on windows 10, UI Automation support is complete enough to provide access to Microsoft Word documents almost on par with NvDA's existing Microsoft Word support, with the added advantage that responsiveness is majorly increased.
```
Comment (leonardder) [inline user_docs/en/userGuide.t2t]: What do you mean with "on par"? I assume it's a typo? Also, note the lower case v in NvDA's

File: user_docs/en/userGuide.t2t [hunk 5]
```diff
@@ -1603,6 +1603,26 @@ This category contains the following options:
...
+However, There may be some information which is either not exposed, or exposed incorrectly in some versions of Microsoft Office, which means this UI automation support cannot always be relied upon.
+We still do not recommend that  the majority of users turn this on by default, though we do welcome users of Office 2016/365 to test this feature and provide feedback.
+
+==== Caret move timeout (in MS) ====
+This option allows you to configure the number of mili seconds NVDA will wait for the caret (insertion point) to move in editable text controls.
+If you find that NVDA seems to be incorrectly tracking the caret E.g. it seems to be always one character behind or is repeating lines, then you may wish to try increasing this value.
+
+==== Debug logging categories ====
+The checkboxes in this group allow you to enable specific categories of debug messages in NVDA's log.
```
Comment (leonardder) [inline user_docs/en/userGuide.t2t]: This is now a list instead of a group, though the list has its own group

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
The code looks ok, just some user guide things

**Comment by DrSooom** [comment]
Describing hwIo, audioDucking, gui, louis and timeSinceInput in the User Guide would be nice.

**Comment by leonardder** [comment]
@DrSooom: I don't think it will add anything useful. These are debug flags that should usually only be set when a developer asks you. If a user doesn't know what they mean, it is probably not necessary for that user to consider checking them.

File: user_docs/en/userGuide.t2t (hunk 4 above)
Comment (feerrenrut) [reply user_docs/en/userGuide.t2t]: @leonardder "on par" is used in golf, and at least quite commonly in Australia, apparently comes from Latin to mean equal to. As a suggestion, the phrase "feature parity" might be preferable.

File: source/_UIAHandler.py [hunk 6]
```diff
@@ -377,21 +374,7 @@ def _isUIAWindowHelper(self,hwnd):
...
-					log.debugWarning("Unknown application using MS Word document control: %s"%appModule.appName)
-					return True
-				try:
-					versionMajor,versionMinor,versionBuild,versionPatch=[int(x) for x in appModule.productVersion.split('.')]
-				except Exception as e:
-					log.error("Error parsing version information %s, %s"%(appModule.productVersion,e))
-					return True
-				if (
-					versionMajor<minMSWordUIAVersion.major
-					or versionMajor==minMSWordUIAVersion.major and versionMinor==minMSWordUIAVersion.minor and versionBuild<minMSWordUIAVersion.build
```
Comment (feerrenrut) [inline source/_UIAHandler.py]: I think we should leave in place checks that exclude versions that we know will not work. Though this should be reflected in the user guide; "This setting will not enable UIA in MS Word versions below 16.0.9000"

**Review by LeonarddeR — APPROVED** [review]

**Pull request merged**